AI/BattleAI/AttackPossibility.cpp

```diff
@@ -1,97 +1,191 @@
-/*
- * AttackPossibility.cpp, part of VCMI engine
- *
- * Authors: listed in file AUTHORS in main folder
- *
- * License: GNU General Public License v2.0 or later
- * Full text of license available in license.txt file, in main folder
- *
- */
-#include "StdInc.h"
-#include "AttackPossibility.h"
-
-AttackPossibility::AttackPossibility(BattleHex tile_, const BattleAttackInfo & attack_)
-	: tile(tile_),
-	attack(attack_)
-{
-}
-
-
-int64_t AttackPossibility::damageDiff() const
-{
-	//TODO: use target priority from HypotheticBattle
-	const auto dealtDmgValue = damageDealt;
-	const auto receivedDmgValue = damageReceived;
-
-	int64_t diff = 0;
-
-	//friendly fire or not
-	if(attack.attacker->unitSide() == attack.defender->unitSide())
-		diff = -dealtDmgValue - receivedDmgValue;
-	else
-		diff = dealtDmgValue - receivedDmgValue;
-
-	//mind control
-	auto actualSide = getCbc()->playerToSide(getCbc()->battleGetOwner(attack.attacker));
-	if(actualSide && actualSide.get() != attack.attacker->unitSide())
-		diff = -diff;
-	return diff;
-}
-
-int64_t AttackPossibility::attackValue() const
-{
-	return damageDiff() + tacticImpact;
-}
-
-AttackPossibility AttackPossibility::evaluate(const BattleAttackInfo & attackInfo, BattleHex hex)
-{
-	const std::string cachingStringBlocksRetaliation = "type_BLOCKS_RETALIATION";
-	static const auto selectorBlocksRetaliation = Selector::type(Bonus::BLOCKS_RETALIATION);
-
-	const bool counterAttacksBlocked = attackInfo.attacker->hasBonus(selectorBlocksRetaliation, cachingStringBlocksRetaliation);
-
-	AttackPossibility ap(hex, attackInfo);
-
-	ap.attackerState = attackInfo.attacker->acquireState();
-
-	const int totalAttacks = ap.attackerState->getTotalAttacks(attackInfo.shooting);
-
-	if(!attackInfo.shooting)
-		ap.attackerState->setPosition(hex);
-
-	auto defenderState = attackInfo.defender->acquireState();
-	ap.affectedUnits.push_back(defenderState);
-
-	for(int i = 0; i < totalAttacks; i++)
+/*
+ * AttackPossibility.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+#include "StdInc.h"
+#include "AttackPossibility.h"
+#include "../../lib/CStack.h"//todo: remove
+
+AttackPossibility::AttackPossibility(BattleHex from, BattleHex dest, const BattleAttackInfo & attack)
+	: from(from), dest(dest), attack(attack)
+{
+}
+
+int64_t AttackPossibility::damageDiff() const
+{
+	return damageDealt + damageReceived + collateralDamage + shootersBlockedDmg;

```

Comment:
Shouldn't damage received be negative? The more damage we receive, the lower function value.

Same goes with collateral damage.